### PR TITLE
wasm: pass PluginHandle to stream contexts.

### PR DESCRIPTION
**Pull request by mathetake**
Previously, stream contexts (i.e. filter instances) are not holding shared ptrs to Wasm plugins which take ownership of Wasm VMs, and only holding raw pointers to Wasm VMs. If Wasm plugins (i.e. factory) die earlier than stream contexts during ECDS updates, then the left stream contexts will try using invalidated pointers to Wasm VM and this results in crashes.
So this commit passes the PluginHandle to all the stream contexts which the plugin generates, so the VM must outlive all stream contexts by which it is used.

Counter part in Proxy-Wasm C++ Host: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/167
Resolves https://github.com/istio/istio/issues/33091/ 
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

EDIT: I hope this commit https://github.com/envoyproxy/envoy/pull/16795/commits/412081d5be099369d57ae2e92d8480e7425d43d7 explains the bug well.

### CODE REVIEW

**Comment by mathetake** [comment]
OK we need to pass the handle to ContexBase in C++ host so that VMs outlives not only Context (in Envoy) but also ContextBase (in Proxy-Wasm C++ host).

**Comment by mathetake** [comment]
tests will follow after #16430

**Comment by alyssawilk** [comment]
I think Lizan's out for a bit, so @PiotrSikora can you do first pass, and tag Matt or I when it's ready for merge?

**Review by lizan — APPROVED** [review]

**Comment by lizan** [comment]
@PiotrSikora still need a non-Tetrands review from you

**Comment by PiotrSikora** [comment]
> @PiotrSikora still need a non-Tetrands review from you

Sorry, I spent way too much time doing archeology work on this issue trying to find a regression, and then looking for an alternative solution. I should have timed out on this much sooner, I'll wrap this within next 24 hours.

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -965,8 +965,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_name = "WebAssembly for Proxies (C++ host implementation)",
         project_desc = "WebAssembly for Proxies (C++ host implementation)",
         project_url = "https://github.com/proxy-wasm/proxy-wasm-cpp-host",
-        version = "605ee8a0eb78127e81f53bcc3f0b7ec983fb65c2",
-        sha256 = "7996d1d34ca0cae4079dc2dbe9b8d51ec6db5cb78ab93533bb8cd31ef970b84a",
+        version = "5bc556e5c8e829d59ad7282c0465f610bce01382",
+        sha256 = "3fa66ff992f7a35bd90cb34852fae4a77ff4524e26e1494cadcd23dd50ed944b",
```
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:969]: Please update to the merged commit.

File: source/extensions/common/wasm/wasm.h [hunk 2]
```diff
@@ -137,25 +137,32 @@ class WasmHandle : public WasmHandleBase, public ThreadLocal::ThreadLocalObject
...
 
 private:
+  PluginSharedPtr plugin_;
   WasmHandleSharedPtr wasm_handle_;
-  const uint32_t root_context_id_;
 };
 
 using PluginHandleSharedPtr = std::shared_ptr<PluginHandle>;
 
+class PluginHandleSharedPtrThreadLocalObject : public ThreadLocal::ThreadLocalObject {
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm.h:157]: Nit: I think you could safely drop the `Object` suffix. None of the other classes are named like that in Envoy.

**Review by PiotrSikora — DISMISSED** [review]
Thanks, and sorry for the delay!

File: source/extensions/common/wasm/wasm.h (hunk 2 above)
Comment (mathetake) [reply source/extensions/common/wasm/wasm.h]: done! https://github.com/envoyproxy/envoy/pull/16795/commits/8e236a721cc0a88ec74bdb08ed3bc1d5add12a14

**Comment by mathetake** [comment]
@moderation PTAL for dep!

**Comment by mathetake** [comment]
@lizan I believe we are good to go now

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Review by lizan — APPROVED** [review]

**Pull request merged**